File: scripts/test-cache-coordination.mjs
/** Local runtime stays passive; actual loopback HTTP, typed trace shipment and API ingest. */
import assert from 'node:assert/strict';
import http from 'node:http';
import {once} from 'node:events';
import {spawnSync} from 'node:child_process';
import {createOllamaAdapter} from '../src/shared/ai/providers/local-ollama.js';
import {observeLocalPrefix,createLocalPrefixObserver} from '../src/shared/ai/cache-coordination.js';
import {shortenValue,setTracingEnabled,note,flushTrace} from '../src/shared/trace.js';
import {rememberDiagnostic,recentDiagnostic,clearRecentDiagnostics} from '../src/background/ai/recent-diagnostics.js';
import {formatRequestDiagnostic} from '../src/shared/ai/diagnostic-view.js';
const calls=[],events=[],wire=[];
const server=http.createServer(async(req,res)=>{
 let text='';for await(const c of req)text+=c;calls.push(JSON.parse(text));
 res.writeHead(200,{'content-type':'application/x-ndjson'});
 res.end(JSON.stringify({message:{content:'<<TP_P0:Hello>>'}})+'\n'+JSON.stringify({done:true,done_reason:'stop',prompt_eval_count:3000,eval_count:10})+'\n');
});
server.listen(0,'127.0.0.1');await once(server,'listening');
const url=`http://127.0.0.1:${server.address().port}`,layout={staticPrefixSha256:'a'.repeat(64),targetLang:'th',sourceLang:'en'};
let final;
try{
 const adapter=createOllamaAdapter({baseUrl:url});
 const request={model:'fixture-local',messages:[{role:'system',content:'STYLE_UNCHANGED'},{role:'user',content:'<<TP_P0:Hello>>'}],outputTokens:512,thinkingMode:'off',thinkingCapability:{supported:true,control:'boolean'},contextTokens:12288};
 const ctx={expectedIds:['P0'],cacheContext:layout,cacheRevision:'digest1',trace:(_event,v)=>events.push(v),wireTrace:(event,v)=>wire.push([event,v])};
 const first=await adapter.generate(request,ctx),second=await adapter.generate(request,ctx);
 assert.equal(calls.length,2);assert.deepEqual(calls[0],calls[1]);assert.deepEqual(calls[0],adapter.payload(request));
 assert.equal(first.cacheCoordination.mode,'runtime_managed');assert.equal(first.cacheCoordination.role,'leader');
 assert.equal(second.cacheCoordination.role,'leader');assert.equal(second.cacheCoordination.groupId,first.cacheCoordination.groupId);
 assert.equal(second.cacheCoordination.previousCompletions,1);assert.equal(second.cacheCoordination.waitMs,0);
 assert.equal(second.cacheCoordination.cacheStatus,'not_reported');assert.equal(second.cacheCoordination.cachedInputTokens,null);
 assert.equal(calls[0].think,false);assert.equal(calls[0].options.num_ctx,12288);assert.equal(calls[0].options.temperature,.2);
 const japanese=await adapter.generate(request,{...ctx,cacheContext:{...layout,targetLang:'ja',staticPrefixSha256:'b'.repeat(64)}});
 assert.notEqual(japanese.cacheCoordination.groupId,first.cacheCoordination.groupId);
 const back=await adapter.generate(request,ctx);assert.equal(back.cacheCoordination.groupId,first.cacheCoordination.groupId);assert.equal(back.cacheCoordination.role,'leader');
 const changed=await adapter.generate({...request,model:'other'},ctx);assert.notEqual(changed.cacheCoordination.groupId,back.cacheCoordination.groupId);
 const probe=await adapter.generate(request,{expectedIds:['P0']});assert.equal(probe.cacheCoordination,undefined,'discovery probe must not count as translation/cache leader');
 assert.equal(calls.length,6,'no extra warmup/discovery requests');
 const controller=new AbortController();controller.abort(new DOMException('Cancelled','AbortError'));
 await assert.rejects(adapter.generate(request,{...ctx,signal:controller.signal}),err=>err.cacheCoordination.requestDispatched===false);
 assert.equal(calls.length,6,'cancelled before dispatch remains not sent');
 assert.ok(events.some(e=>e.releaseReason==='leader_cancelled'));
 assert.ok(events.filter(e=>e.schema==='tp.cache_coordination/1').every(e=>e.waitMs===0&&e.providerCacheTtlMs===null));
 const args={url,model:'x',payload:{},layout};
 const one=await observeLocalPrefix({...args,headers:{Authorization:'PRIVATE_ACCOUNT'}}),two=await observeLocalPrefix({...args,headers:{Authorization:'PRIVATE_OTHER'}});
 assert.notEqual(one.snapshot().groupId,two.snapshot().groupId);
 one.dispatched();final=one.finish({cachedInputTokens:123,inputTokens:3000},true);
 assert.equal(final.cacheStatus,'reported_hit');assert.equal(final.providerCacheReady,null);
 assert.ok(wire.some(([event,v])=>event==='providerRequest'&&v.cacheCoordination));
}finally{server.closeAllConnections();server.close();await once(server,'close');}
// One sanitizer shared with both trace paths. Counterfeit readiness and secrets removed.
const dirty={...final,providerCacheReady:true,providerCacheTtlMs:100,operationId:'ai:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',api_key:'PRIVATE',prompt:'PRIVATE',endpoint:'PRIVATE'};
const safe=shortenValue(dirty);assert.equal(safe.providerCacheReady,null);assert.equal(safe.providerCacheTtlMs,null);assert.doesNotMatch(JSON.stringify(safe),/PRIVATE/);assert.ok(Object.keys(safe).length>20);
const python=spawnSync('python',['-c',`import sys,json;sys.path.insert(0,'api');from backend.diagnostic_schema import sanitize_cache_coordination;print(json.dumps(sanitize_cache_coordination(json.load(sys.stdin))))`],{input:JSON.stringify(dirty),encoding:'utf8'});
assert.equal(python.status,0,python.stderr);assert.deepEqual(JSON.parse(python.stdout),safe,'JS/Python schema parity');
rememberDiagnostic({provider:'huggingface',model:'x',operationId:'ai:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',coordination:{...safe,waitMs:210},result:{schema:'tp.audit/1',event:'translation_result',resultStatus:'complete_at_contract_boundary',actualInput:3000,actualOutput:10,cachedInput:123}});
const row=recentDiagnostic('huggingface','x');assert.equal(row.coordination.waitMs,210);
assert.match(formatRequestDiagnostic(row),/Cache wait: 0.21 s/);assert.doesNotMatch(formatRequestDiagnostic(row),/PRIVATE|groupId|providerCacheReady|unknown/);clearRecentDiagnostics();
// Real JS shipment -> actual ASGI route -> actual temp JSONL; duplicate ACK safety.
const originalFetch=globalThis.fetch,shipments=[];
try{
 globalThis.fetch=async(_url,init)=>{shipments.push(JSON.parse(init.body));return new Response('{"ok":true}',{status:200});};
 setTracingEnabled(true,()=> 'http://fixture','compact','cache-fixture');
 note('fixture','cache-proof',dirty,'tcache123456');await flushTrace();setTracingEnabled(false);
 assert.equal(shipments.length,1);
 const child=spawnSync('python',['-c',String.raw`
import sys,json,tempfile,os,asyncio
from pathlib import Path
sys.path.insert(0,'api')
with tempfile.TemporaryDirectory() as temp:
 os.environ.update(TP_TRACE='1',TP_TRACE_DIR=temp,TP_LOG_DIR=temp)
 from backend import trace
 from backend.api.routes.logs import router
 from fastapi import FastAPI
 import httpx
 trace._ROOT=Path(temp);app=FastAPI();app.include_router(router)
 payload=json.load(sys.stdin);payload['traceSession']=trace.session_id()
 async def run():
  async with httpx.AsyncClient(transport=httpx.ASGITransport(app=app),base_url='http://fixture') as c:
   a=await c.post('/v1/trace',json=payload);b=await c.post('/v1/trace',json=payload)
   assert a.status_code==200 and b.status_code==200,(a.text,b.text)
   assert b.json()['written']==0,b.text
 asyncio.run(run());trace.flush()
 text='\n'.join(p.read_text() for p in Path(temp).glob('*.jsonl'))
 rows=[json.loads(s)['d'] for s in text.splitlines() if s and json.loads(s).get('fn')=='cache-proof']
 assert len(rows)==1,rows
 d=rows[0];assert d['schema']=='tp.cache_coordination/1'
 assert d['cachedInputTokens']==123 and d['cacheStatus']=='reported_hit'
 assert d['providerCacheReady'] is None and d['missReason']=='unknown'
 assert d['providerCacheTtlMs'] is None and d['coordinationPolicy']=='observe_no_wait'
 assert d['leaderLeaseState']=='released' and d['observationRecorded'] is True
 assert d['operationId']=='ai:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' and d['staticPrefixSha256']=='a'*64,d
 assert len(d)>20 and 'PRIVATE' not in text
 print('PASS complete cache event (>20 fields), HTTP ingestion, deduplicated shipment and secret redaction')
`],{input:JSON.stringify(shipments[0]),encoding:'utf8'});
 assert.equal(child.status,0,child.stderr||child.stdout);console.log(child.stdout.trim());
}finally{setTracingEnabled(false);globalThis.fetch=originalFetch;}
console.log('PASS Local loopback HTTP / model-language-account namespaces / cancellation / no extra requests / compact UI / cross-runtime trace');

// Independent statistics/active leases, no timers, late-result fences and rotation.
let now=0;const observer=createLocalPrefixObserver({clock:()=>now,leaseMs:100,maxGroups:1,maxActive:2});
const config={url:'http://localhost:11434',model:'x',payload:{},layout};
const a=await observer(config);a.dispatched();
const other=await observer({...config,model:'y'});
const b=await observer(config);b.dispatched();
assert.equal(b.snapshot().role,'observer');assert.equal(a.snapshot().leaderCoordinationId,b.snapshot().leaderCoordinationId);
a.finish({inputTokens:100,cachedInputTokens:10},true);assert.equal(a.snapshot().observationRecorded,false);
b.finish({inputTokens:100,cachedInputTokens:0},true);
const c=await observer(config);c.dispatched();assert.equal(c.snapshot().reason,'next_request');
now=101;const d=await observer(config);d.dispatched();assert.equal(d.snapshot().reason,'lease_expired');
d.finish({inputTokens:100,cachedInputTokens:5},true);c.finish({inputTokens:100,cachedInputTokens:0},true);
assert.equal(c.snapshot().releaseReason,'lease_superseded');assert.equal(c.snapshot().latestObservationApplied,false);
now=9999999;const e=await observer(config);assert.equal(e.snapshot().previousCacheHit,true);assert.equal(e.snapshot().retentionMs,null);
assert.equal(e.snapshot().waitMs,0);other.finish({},false);e.finish({},false);
console.log('PASS Local observation LRU independent of active lease / logical expiry / request-order fence / no wait');

File: scripts/test-cache-coordination.py
"""No-wait leases and bounded observations; no external model calls."""
import os, sys, time, threading, unittest
from concurrent.futures import ThreadPoolExecutor
from pathlib import Path
from unittest.mock import patch
sys.path.insert(0, str(Path(__file__).resolve().parents[1] / 'api'))
from backend.ai import cache_coordination as mod
from backend.ai.clients.base import ProviderGenerationCancelled
from backend.diagnostic_schema import sanitize_cache_coordination

BASE=dict(provider='huggingface',model='fixture-model',endpoint='https://router.huggingface.co/v1',
          account='PRIVATE_A',prefix='a'*64,target='th',source='en')

class Tests(unittest.TestCase):
 def setUp(self):
  self.events=[];self.p=patch.object(mod,'emit',side_effect=self.events.append);self.p.start();self.addCleanup(self.p.stop)
 def test_no_wait_for_active_leader(self):
  c=mod.PrefixCoordinator();a=c.acquire(**BASE)
  with ThreadPoolExecutor(6) as p:
   futures=[p.submit(c.acquire,**BASE) for _ in range(20)]
   items=[f.result(timeout=.5) for f in futures]
  self.assertTrue(all(i.data['role']=='observer' and i.data['waitMs']==0 for i in items))
  self.assertFalse(a.finished);self.assertEqual(len(c._leaders),1)
  self.assertTrue(all(e['phase']!='waiting' for e in self.events))
 def test_next_real_request_becomes_leader_on_terminal(self):
  c=mod.PrefixCoordinator();a=c.acquire(**BASE);a.dispatched();a.finish(complete=True,cached=10,input_tokens=20)
  b=c.acquire(**BASE)
  self.assertEqual(b.key,a.key);self.assertEqual(b.data['role'],'leader');self.assertEqual(b.data['reason'],'next_request')
  self.assertTrue(b.data['previousCacheHit']);self.assertEqual(b.data['previousCompletions'],1)
  self.assertNotEqual(a.data['leaderCoordinationId'],b.data['leaderCoordinationId'])
 def test_failure_and_cancel_release_without_retry(self):
  for reason in ['failed','cancelled']:
   c=mod.PrefixCoordinator();a=c.acquire(**BASE);a.dispatched();a.finish(reason=reason)
   self.assertEqual(a.data['releaseReason'],'leader_'+reason)
   b=c.acquire(**BASE);self.assertEqual(b.data['role'],'leader');self.assertEqual(b.data['waitMs'],0)
 def test_cancel_before_acquire(self):
  c=mod.PrefixCoordinator()
  with self.assertRaises(ProviderGenerationCancelled):c.acquire(**BASE,cancel_check=lambda:True)
  self.assertEqual(len(c._leaders),0);self.assertEqual(len(c._observations),0)
 def test_lease_expiry_and_fenced_late_completion(self):
  now=[0.];c=mod.PrefixCoordinator(clock=lambda:now[0],lease_ms=100)
  a=c.acquire(**BASE);a.dispatched();now[0]=.11;b=c.acquire(**BASE);b.dispatched()
  self.assertEqual(b.data['reason'],'lease_expired');b.finish(complete=True,cached=0)
  a.finish(complete=True,cached=25)
  self.assertEqual(a.data['releaseReason'],'lease_superseded');self.assertFalse(a.data['latestObservationApplied'])
  d=c.acquire(**BASE);self.assertFalse(d.data['previousCacheHit']);self.assertEqual(d.data['previousCompletions'],2)
 def test_observer_cannot_release_current_leader(self):
  c=mod.PrefixCoordinator();a=c.acquire(**BASE);b=c.acquire(**BASE);b.dispatched();b.finish(complete=True,cached=10)
  self.assertIs(c._leaders[a.key],a.leader);a.dispatched();a.finish(complete=True,cached=0)
  self.assertFalse(a.data['latestObservationApplied']);self.assertTrue(c.acquire(**BASE).data['previousCacheHit'])
 def test_eviction_does_not_evict_active_leases_or_restore_stale_stats(self):
  c=mod.PrefixCoordinator(max_groups=1,max_active=4)
  a=c.acquire(**BASE);a.dispatched();other=c.acquire(**{**BASE,'account':'OTHER'})
  b=c.acquire(**BASE);b.dispatched()
  self.assertIs(a.leader,b.leader);self.assertEqual(b.data['role'],'observer')
  a.finish(complete=True,cached=10);self.assertFalse(a.data['observationRecorded'])
  b.finish(complete=True,cached=0);self.assertTrue(b.data['observationRecorded'])
  self.assertIs(c._leaders[other.key],other.leader)
 def test_capacity_never_blocks_translation_or_kills_leader(self):
  c=mod.PrefixCoordinator(max_groups=1,max_active=1)
  a=c.acquire(**BASE);b=c.acquire(**{**BASE,'model':'OTHER'})
  self.assertEqual(b.data['reason'],'leader_capacity');self.assertIsNone(b.leader)
  b.dispatched();b.finish(complete=True,cached=0)
  self.assertIs(c._leaders[a.key],a.leader)
 def test_idle_statistics_are_not_provider_ttl(self):
  now=[0.];c=mod.PrefixCoordinator(clock=lambda:now[0]);a=c.acquire(**BASE);a.dispatched();a.finish(complete=True,cached=5)
  now[0]=6000;v=c.acquire(**BASE)
  self.assertTrue(v.data['reusedGroup']);self.assertEqual(v.data['role'],'leader');self.assertEqual(v.data['waitMs'],0)
  self.assertIsNone(v.data['retentionMs']);self.assertIsNone(v.data['providerCacheTtlMs'])
  self.assertEqual(v.snapshot()['previousObservationAgeMs'],6000000)
 def test_zero_unknown_hit_and_invalid_usage(self):
  for cached,input_t,expected in [(None,100,'not_reported'),(0,100,'reported_zero'),(50,100,'reported_hit'),(150,100,'not_reported'),(True,100,'not_reported')]:
   c=mod.PrefixCoordinator();v=c.acquire(**BASE);v.dispatched();v.finish(complete=True,cached=cached,input_tokens=input_t)
   self.assertEqual(v.data['cacheStatus'],expected);self.assertIsNone(v.data['providerCacheReady'])
   self.assertIsNone(v.data['providerCacheTtlMs']);self.assertEqual(v.data['missReason'],'unknown')
 def test_unknown_newer_result_does_not_keep_old_hit(self):
  c=mod.PrefixCoordinator();a=c.acquire(**BASE);a.dispatched();a.finish(complete=True,cached=20)
  b=c.acquire(**BASE);b.dispatched();b.finish(complete=True)
  self.assertIsNone(c.acquire(**BASE).data['previousCacheHit'])
 def test_finish_idempotent_and_unattempted_usage_not_learned(self):
  c=mod.PrefixCoordinator();a=c.acquire(**BASE);a.finish(complete=True,cached=9)
  self.assertFalse(a.data['observationRecorded']);self.assertEqual(c.acquire(**BASE).data['previousCompletions'],0)
  b=c.acquire(**BASE);b.dispatched();x=b.finish(complete=True,cached=0);self.assertEqual(x,b.finish(complete=True,cached=5))
 def test_multiuser_multimodel_language_isolation_and_switchback(self):
  c=mod.PrefixCoordinator();keys=set()
  for account in range(30):
   for model in range(4):
    for lang in ['en','th','ja']:
     v=c.acquire(**{**BASE,'account':str(account),'model':str(model),'target':lang});v.dispatched();v.finish(complete=True,cached=0)
     keys.add(v.key)
  self.assertEqual(len(keys),360);self.assertEqual(len(c._observations),360);self.assertEqual(len(c._leaders),0)
  a=c.acquire(**BASE);a.finish()
  for field,value in [('account','PRIVATE_B'),('model','OTHER'),('endpoint','https://other.example'),('prefix','b'*64),('target','ja'),('source','ja'),('revision','r2'),('thinking','on'),('has_image',True),('response_schema',{'type':'object'})]:
   v=c.acquire(**{**BASE,field:value});self.assertNotEqual(v.key,a.key);v.finish()
  self.assertEqual(c.acquire(**BASE).key,a.key)
 def test_processes_are_independent_not_distributed(self):
  a,b=mod.PrefixCoordinator(),mod.PrefixCoordinator()
  self.assertNotEqual(a.acquire(**BASE).key,b.acquire(**BASE).key)
 def test_config_modes_and_legacy_wait_do_not_add_delay(self):
  with patch.dict(os.environ,{'TP_PROMPT_CACHE':'auto','TP_PROMPT_CACHE_COORDINATION':'auto','TP_PROMPT_CACHE_WAIT_MS':'10000'}):
   self.assertEqual(mod._mode('huggingface','x',BASE['endpoint']),'observe_only')
   self.assertEqual(mod._mode('huggingface','x','https://router.huggingface.co.evil/v1'),'unsupported')
   self.assertEqual(mod._mode('ollama','x','http://localhost:11434'),'runtime_managed')
  with patch.dict(os.environ,{'TP_PROMPT_CACHE_COORDINATION':'off'}):self.assertEqual(mod._mode('huggingface','x',BASE['endpoint']),'disabled')
  for mode in ['runtime_managed','disabled','unsupported']:
   c=mod.PrefixCoordinator();a=c.acquire(**BASE,mode=mode);b=c.acquire(**BASE,mode=mode);self.assertEqual(b.data['waitMs'],0)
  self.assertEqual(mod.PrefixCoordinator().acquire(**{**BASE,'prefix':''}).data['role'],'bypass')
 def test_sanitizer_cannot_claim_provider_expiry_readiness_or_expose_secret(self):
  v=mod.PrefixCoordinator().acquire(**BASE).snapshot()
  s=sanitize_cache_coordination({**v,'providerCacheTtlMs':100,'providerCacheReady':True,'api_key':'PRIVATE','prompt':'PRIVATE'})
  self.assertIsNone(s['providerCacheTtlMs']);self.assertIsNone(s['providerCacheReady']);self.assertNotIn('PRIVATE',str(s));self.assertGreater(len(s),30)
 def test_emit_failure_releases_lease(self):
  c=mod.PrefixCoordinator()
  with patch.object(mod,'emit',side_effect=RuntimeError('diagnostic write')):
   with self.assertRaises(RuntimeError):c.acquire(**BASE)
  self.assertEqual(len(c._leaders),0)

if __name__=='__main__':unittest.main()
